pbft/pbft.cpp

```diff
@@ -994,6 +1007,533 @@ pbft::get_high_water_mark()
+            {
+                LOG(error) << "is_valid_viewchange_message - a prepare message has a bad envelope";
+                return false;
+            }
+
+            // does the sequence number, view number and hash match those of the pre prepare
+            if (pbft_msg prepare_msg; !prepare_msg.ParseFromString(prepare_envelope.pbft()) || preprepare_message.sequence() != prepare_msg.sequence() || preprepare_message.view() != prepare_msg.view() || preprepare_message.request_hash() != prepare_msg.request_hash())
+            {
+                LOG(error) << "is_valid_viewchange_message - a prepare message is invalid";
+                return false;
+            }
+
+            senders.insert(prepare_envelope.sender());
+        }
+
+        if (senders.size() < 2 * this->max_faulty_nodes() + 1)
+        {
+            LOG(error) << "is_valid_viewchange_message - there were not enough valid viewchange messages";
+            return false;
+        }
+    }
+    return valid_checkpoint != std::nullopt;
+}
+
+bool
+pbft::get_sequences_and_request_hashes_from_proofs(
+        const pbft_msg& viewchange_msg
+        , std::set<std::pair<uint64_t, std::string>>& sequence_request_pairs) const
+{
+    for (int j{0}; j < viewchange_msg.prepared_proofs_size(); ++j)
+    {
+        // --- add the sequence/hash to a set
+        auto prepared_proof = viewchange_msg.prepared_proofs(j);
+
+        pbft_msg msg;
+        if (bzn_envelope envelope;
+            !envelope.ParseFromString(prepared_proof.pre_prepare())
+            || !this->is_peer(envelope.sender())
+            || !msg.ParseFromString(envelope.pbft()))
+        {
+            return false;
+        }
+        sequence_request_pairs.insert(std::make_pair(msg.sequence(), msg.request_hash()));
+    }
+    return true;
+}
+
+/**
+ * A valid newview message must contain a complete set of pre prepares, that is
+ * there must be one pre prepare for each sequence.
+ * TODO: this method needs a better name
+ * @param newview_msg
+ * @return true if there are no missing pre-prepares based on sequence numbers of each pre-prepare, false otherwise.
+ */
+bool
+pbft::pre_prepares_contiguous(const pbft_msg& newview_msg) const
+{
+    uint64_t minimum_sequence{UINT64_MAX};
+    uint64_t maximum_sequence{0};
+    for (int i{0}; i < newview_msg.pre_prepare_messages_size(); ++i)
+    {
+        const bzn_envelope pre_prepare_envelope{newview_msg.pre_prepare_messages(i)};
+        pbft_msg pre_prepare;
+        pre_prepare.ParseFromString(pre_prepare_envelope.pbft());
+
+        minimum_sequence = pre_prepare.sequence() < minimum_sequence ? pre_prepare.sequence() : minimum_sequence;
+        maximum_sequence = pre_prepare.sequence() > maximum_sequence ? pre_prepare.sequence() : maximum_sequence;
+    }
+    return (maximum_sequence-minimum_sequence) == uint64_t(newview_msg.pre_prepare_messages_size());
```

Comment:
Re wrote pre_prepares_contiguous to walk through the pre prepares and check each sequence number with the previous one. This will catch missing, out of order and duplicate sequence numbers.

Empty pre prepares lists are considered to be contiguous.